lib/launch.c

```diff
@@ -286,8 +290,30 @@ pop_n_at_stage3(int n)
 	}
 }
 
+static int
+drop_privileges(const char *user)
+{
+	struct passwd *pw;
+
+	if ((pw = getpwnam(user)) == NULL) {
+		G_LOG(ERR, "%s: failed to call getpwnam() for user %s!\n",
+			__func__, user);
+		return -1;
+	}
+
+	/* Drop privileges. */
+	if (initgroups(user, pw->pw_gid) != 0 || setgid(pw->pw_gid) != 0 ||
+			setuid(pw->pw_uid) != 0) {
+		G_LOG(ERR, "%s: failed to drop privileges for user %s!\n",
+			__func__, user);
+		return -1;
+	}
+
+	return (seteuid(0) == -1 && setegid(0) == -1) ? 0 : -1;
+}
+
 int
-launch_gatekeeper(void)
+launch_gatekeeper(const char *user)
```

Comment:
We don't need to patch the launch library, just use `finalize_stage2()` and update the comment in `main/main.c` on the line `ret = launch_at_stage2(finalize_stage2, NULL);`.